gazebo_plugins/src/gazebo_ros_ray_sensor.cpp

```diff
@@ -0,0 +1,428 @@
+        _sdf->Get<std::string>("radiationType").c_str());
+      return;
+    }
+  }
+
+  if (!_sdf->HasElement("minIntensity")) {
+    impl_->min_intensity_ = 0.0;
+    RCLCPP_DEBUG(impl_->rosnode_->get_logger(), "missing <minIntensity>, defaults to %f",
+      impl_->min_intensity_);
+  } else {
+    impl_->min_intensity_ = _sdf->Get<double>("minIntensity");
+  }
+
+
+  // Create gazebo tranport node and subscribe to sensor's laser scan
+  impl_->gazebo_node_ = boost::make_shared<gazebo::transport::Node>();
+  impl_->gazebo_node_->Init(_parent->WorldName());
+  // TODO(ironmig): use lazy publisher to only process laser data when output has a subscriber
+  impl_->SubscribeGazeboLaserScan();
+}
+
+template<typename T>
+void GazeboRosRaySensorPrivate::SetParams(T sensor)
+{
+  minAngle = sensor->AngleMin().Radian();
+  maxAngle = sensor->AngleMax().Radian();
+  verticalMaxAngle = sensor->VerticalAngleMax().Radian();
+  verticalMinAngle = sensor->VerticalAngleMin().Radian();
+  rayCount = sensor->RayCount();
+  rangeCount = sensor->RangeCount();
+  verticalRayCount = sensor->VerticalRayCount();
+  verticalRangeCount = sensor->VerticalRangeCount();
+  yDiff = maxAngle - minAngle;
+  pDiff = verticalMaxAngle - verticalMinAngle;
+  rangeMin = sensor->RangeMin();
+  rangeMax = sensor->RangeMax();
+  sensor_topic_ = sensor->Topic();
+}
+
+void GazeboRosRaySensorPrivate::SubscribeGazeboLaserScan()
+{
+  if (pub_.type() == typeid(LaserScanPub)) {
+    laser_scan_sub_ = gazebo_node_->Subscribe(sensor_topic_,
+        &GazeboRosRaySensorPrivate::PublishLaserScan, this);
+  } else if (pub_.type() == typeid(PointCloudPub)) {
+    laser_scan_sub_ = gazebo_node_->Subscribe(sensor_topic_,
+        &GazeboRosRaySensorPrivate::PublishPointCloud, this);
+  } else if (pub_.type() == typeid(PointCloud2Pub)) {
+    laser_scan_sub_ = gazebo_node_->Subscribe(sensor_topic_,
+        &GazeboRosRaySensorPrivate::PublishPointCloud2, this);
+  } else if (pub_.type() == typeid(RangePub)) {
+    laser_scan_sub_ = gazebo_node_->Subscribe(sensor_topic_,
+        &GazeboRosRaySensorPrivate::PublishRange, this);
+  } else {
+    RCLCPP_ERROR(rosnode_->get_logger(), "Publisher is an invalid type. This is an internal bug.");
+  }
+}
+
+void GazeboRosRaySensorPrivate::PublishLaserScan(ConstLaserScanStampedPtr & _msg)
+{
+  sensor_msgs::msg::LaserScan ls;
+  ls.header.frame_id = frame_name_;
+  ls.header.stamp.sec = _msg->time().sec();
+  ls.header.stamp.nanosec = _msg->time().nsec();
+  ls.angle_min = minAngle;
+  ls.angle_max = maxAngle;
+  ls.angle_increment = _msg->scan().angle_step();
+  ls.time_increment = 0;
+  ls.scan_time = 0;
+  ls.range_min = _msg->scan().range_min();
+  ls.range_max = _msg->scan().range_max();
+
+  // If there are multiple vertical beams, use the one in the middle
+  size_t start = (verticalRangeCount / 2) * rangeCount;
+
+  // Copy ranges and intensities over
+  ls.ranges.resize(rangeCount);
```

Comment:
Why not use `_msg->scan().ranges_size()`?

Reply:
I think this would be a problem if there is greater than one vertical ray as this is just a planar scan. rangeCount is the number of horizontal rays, where as _msg->scan().ranges_size() is all the rays